lib/Conversion/FIRRTLToHW/LowerToHW.cpp

```diff
@@ -892,6 +914,74 @@ FIRRTLModuleLowering::lowerModule(FModuleOp oldModule, Block *topLevelModule,
   if (annos.removeAnnotation(verifBBClass))
     newModule->setAttr("firrtl.extract.cover.extra", builder.getUnitAttr());
 
+  bool failed = false;
+  // Remove ForceNameAnnotations by generating verilogNames on instances.
+  annos.removeAnnotations([&](Annotation anno) {
+    if (!anno.isClass(forceNameClass))
+      return false;
+
+    auto sym = anno.getMember<FlatSymbolRefAttr>("circt.nonlocal");
+    // This must be a non-local annotation due to how the Chisel API is
+    // implemented.
+    //
+    // TODO: handle this in some sensible way based on what the SFC does with
+    // a local annotation.
+    if (!sym) {
+      auto diag = oldModule.emitOpError()
+                  << "contains a '" << forceNameClass
+                  << "' that is not a non-local annotation";
+      diag.attachNote() << "the erroneous annotation is '" << anno.getDict()
+                        << "'\n";
+      failed = true;
+      return false;
+    }
+
+    auto nla = loweringState.lookupNLA(sym);
+    // The non-local anchor must exist.
+    //
+    // TODO: handle this with annotation verification.
+    if (!nla) {
+      auto diag = oldModule.emitOpError()
+                  << "contains a '" << forceNameClass
+                  << "' whose non-local symbol, '" << sym
+                  << "' does not exist in the circuit";
+      diag.attachNote() << "the erroneous annotation is '" << anno.getDict();
+      failed = true;
+      return false;
+    }
+
+    // Add the forced name to global state (keyed by a pseudo-inner name ref).
+    // Error out if this key is alredy in use.
+    //
+    // TODO: this error behavior can be relaxed to always overwrite with the
+    // new forced name (the bug-compatible behavior of the Chisel
+    // implementation) or fixed to duplicate modules such that the naming can
+    // be applied.
+    auto mod = nla.modpath()
+                   .getValue()
+                   .take_back(2)[0]
+                   .cast<FlatSymbolRefAttr>()
+                   .getAttr();
+    auto inst = nla.namepath().getValue().take_back(2)[0];
```

Comment:
This is interesting, so it needs only the last instance in the instance path, which will be renamed.
Another obvious observation here is, we cannot construct a local annotation with a `nla`, that is the `namepath` must have at least two entries. Maybe we can add this to the verifier for the `nla`

Reply:
Yep. It's due to a quirk in how the NLA looks for `ForceNameAnnotation`.  The actual annotation lives on the _module_ (`Bar` below), when the information is actually on the _instance_.

```mlir
firrtl.module @Bar() attributes {annotations = [{circt.nonlocal = @nla, ...}]}
firrtl.module @Foo() {
  firrtl.instance bar {annotations = [circt.nonlocal = @nla]}
}
```

If we change this to put the annotation payload on the instance, then this may be cleaner.  Lenharth and I went back and forth on this.  I'm not sure it's super important other than that there is a convention that we always use.

> that is the namepath must have at least two entries

Yes!  This actually came up with me hacking on module inlining.  If you inline a module you can wind up with the NLA being size 1:


```mlir
```